Cancellation and timeouts tutorial review (#4618)
Kotlin/kotlinx.coroutines: docs/topics/coroutines-cancellation.md

You can call the `cancel()` function manually, or it can be invoked automatically through cancellation propagation when a parent coroutine is canceled.

When a coroutine is canceled, it throws a [`CancellationException`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-cancellation-exception/) the next time it checks for cancellation.

> The [Suspension points and cancellation](#suspension-points-and-cancellation) section covers precisely when this happens, but for now it is enough to know that all suspend functions in the `kotlinx.coroutines` library, such as `delay()` and `awaitCancellation()` explained below, check for cancellation if they suspend.

Reviewer: As a general comment: 
The docs purposefully avoid similar phrases like: explained below — these are quite fragile, because the surrounding structure can change, and don't add much to the reader, since below is not exact. (a link is exact)

I think the original is more concise and better captures the overall style, which should be something the readers are already used to. The information you are proposing is a great addition I added a [suggestion](https://github.com/Kotlin/kotlinx.coroutines/pull/4618/changes#r3511386564) how I think we could add it. 👍

Author: This potential hedging interferes with the flow of the text. There isn't much value in a discipline to fix an unlikely potential problem if it's hurting the immediate value. Moreover, the surrounding structure is unlikely to change. These docs will probably never be updated in future given where the world is headed. Even without the world, these docs saw little attention in past. Re your suggestion - please read my comment at the same line as your suggestion.

Reviewer: I still think a link is more exact than "below". If a reader wants to read more about something now, we should point them to the relevant section.

> [`awaitCancellation()`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/await-cancellation.html) function suspends a coroutine until it's canceled. It is equivalent to `delay(Duration.INFINITE)`.

Reviewer: ```suggestion
Suspending functions in the `kotlinx.coroutines` library, such as the `delay()` function, check for cancellation when they suspend.

> For more information about how and when this happens, see [Suspension points and cancellation](#suspension-points-and-cancellation).
>
{style="tip"}

You can use the [`awaitCancellation()`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/await-cancellation.html) function to suspend a coroutine until it's canceled. It's equivalent to calling `delay(Duration.INFINITE)`.
```

Reviewer: I'd suggest using a tip to point readers to more information — it helps it pop and it's the same as saying "The user doesn't need to know this now, and they can read more now if they prefer"

Author: (Please read my comment at the same line)

Reviewer: I'm raising this again because I agree with your sentiment, but this line doesn't fit this though:

> The [Suspension points and cancellation](https://github.com/Kotlin/kotlinx.coroutines/blob/c74267417d125cb61bcd2be189d1b68cd48fab46/docs/topics/cancellation-and-timeouts.md#suspension-points-and-cancellation) section covers precisely when this happens, but for now it is enough to know that all suspend functions in the kotlinx.coroutines library, such as delay() and awaitCancellation() explained below, check for cancellation if they suspend.

This parses really strangely: It's too long, so it's difficult to follow. It also mixes in information that's neither described nor necessary to understand the example or the section.

We should bring this into focus, as this is the important part to understand the example and section:

> Suspending functions in the `kotlinx.coroutines` library, such as the `delay()` function, check for cancellation when they suspend.

And this is a link where the reader can learn more (at this point, optional content, which is the perfect use case for a tip): 

> For more information about how and when this happens, see [Suspension points and cancellation](#suspension-points-and-cancellation).

You can call the `cancel()` function manually, or it can be invoked automatically through cancellation propagation when a parent coroutine is canceled.

When a coroutine is canceled, it throws a [`CancellationException`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-cancellation-exception/) the next time it checks for cancellation.

> The [Suspension points and cancellation](#suspension-points-and-cancellation) section covers precisely when this happens, but for now it is enough to know that all suspend functions in the `kotlinx.coroutines` library, such as `delay()` and `awaitCancellation()` explained below, check for cancellation if they suspend.

Reviewer: Another general comment: the page uses contractions, so we should use them consistently. (it is -> it's | does not -> doesn't etc.)

> suspend fun main() {

Reviewer: I see some consistency issues in hiding `suspend fun main()`, but it's not a big concern, and I think readability is more important, so let's go for it. 👍 

I'm not certain on the `withContext()` part, though 🤔 Wasn't the original intention here to ensure that people see at a glance which dispatcher is used in the examples? (alongside with the examples always specifying a dispatcher explicitly — that's not a concern here) 🤔 
if that's not a big issue, I think we can hide that as well behind sampleStart.

## Make coroutines react to cancellation {id="cancellation-is-cooperative"}

In Kotlin, coroutine cancellation is _cooperative_.

> Coroutines react to cancellation only when they cooperate by [suspending](#suspension-points-and-cancellation) or [checking for cancellation explicitly](#check-for-cancellation-explicitly).

Reviewer: not sure if we have to remove "This means" here. it reinforces that this is a new concept for readers. that said I don't feel strongly about this, I think it works without it as well. why did you decide to remove it? 🤔

Author: Cursive reinforces the newness of the concept. "This means" is hence unnecessary.

> There are standard approaches for making coroutines cancelable, such as periodically calling the [`yield()`](#yield-often-in-non-suspending-code) suspending function in CPU-intensive code.

Reviewer: I find the logic strange here and I much prefer the original. I understand we want to decrease the relevance of ensureActive and isActive, so I suggest removing the link 🤔 — alternatively, we could integrate the yield function in the title and cover it there, and keep the link.

I think this section should serve these functions:
* Explain what cooperative cancellation is as a concept (since there are resources that link here)
* (optional, we can drop this) Let the reader know what they can learn in the subsections

To achieve this I think we have two ways. Either mentioning the subsections briefly in a sentence or add a generic sentence that encompasses all of them like we had before:

> In this section, you can learn how to make coroutines react to cancellation using suspension points, the `yield()` function, explicit cancellation checks, and the `runInterruptible()` function on the JVM.

Reviewer: A related question:
- Do we want to describe yield as adding a suspension point or as an explicit check? 🤔 To me it kind of feels like both, but please correct me if I'm wrong.

I'm asking because if we could integrate it in "Check for cancellation explicitly" it would make this part easier to grasp, but if it's not correct then let's keep it as a separate one.

Author: Yield is both, and it is described as both, so no problem here.

Author: I don't understand your first comment, but the suggested text is not correct, since yield is both.

Reviewer: My issue with this version is that it adds redundant information. 

To make it a bit more focused, I'd propose two options

`Option #1`:
Keep the beginning (except change this part to bring the "react" part into focus

> There are standard approaches for making coroutines react to cancellation, such as periodically calling the [`yield()`](#yield-often-in-non-suspending-code) suspending function in CPU-intensive code.

then remove everything from below this and replace with:

> In this section, you can learn how these approaches work and when to use each one.

`Option #2`:

Remove this sentence entirely: "There are standard approaches for making coroutines react to cancellation...."

and just remove the explicit checking from the short intro, putting the focus on yield() and suspension points.

> In this section, you can learn how to make coroutines react to cancellation using suspension points, the `yield()` function, and the `runInterruptible()` function on the JVM.

On the JVM, you can use [`runInterruptible()`](#interrupt-blocking-code-when-coroutines-are-canceled) with some blocking calls that support thread interruption.

Coroutines that react to cancellation are sometimes referred to as **cancelable coroutines**.

> In this section, you will learn how to create cancelable coroutines.

Reviewer: ```suggestion
In this section, you can learn how to create cancelable coroutines.
```

Reviewer: I think that if there is an issue with the original sentence, then the right approach is to remove it. Replacing "can" with "will" makes this read more like a course/tutorial, so I would strongly advise against it. I have no issue with the original, but I believe it doesn't add much, so feel free to remove.

## Make coroutines react to cancellation {id="cancellation-is-cooperative"}

In Kotlin, coroutine cancellation is _cooperative_.
Coroutines react to cancellation only when they cooperate by [suspending](#suspension-points-and-cancellation) or [checking for cancellation explicitly](#check-for-cancellation-explicitly).
There are standard approaches for making coroutines cancelable, such as periodically calling the [`yield()`](#yield-often-in-non-suspending-code) suspending function in CPU-intensive code.
On the JVM, you can use [`runInterruptible()`](#interrupt-blocking-code-when-coroutines-are-canceled) with some blocking calls that support thread interruption.

> Coroutines that react to cancellation are sometimes referred to as **cancelable coroutines**.

Reviewer: Is there a reason to mention this? I suggest removing this line and modify the next sentence if you think the it sounds like a term 🤔 i.e.: In this section, you can learn how to make coroutines react to cancellation. — but as I said before I think we can safely remove that line as well

>
{style="tip"}

> ### `yield()` often in non-suspending code

Reviewer: ```suggestion
### The `yield()` suspending function
```

>
{style="tip"}

### `yield()` often in non-suspending code

If a coroutine doesn't suspend for a long time, it doesn't stop when it's canceled.
In CPU-intensive computations and in code that rarely reaches a suspension point, call `yield()` periodically to let the current coroutine check for cancellation.

The [`yield()`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/yield.html) function suspends the coroutine, releasing its current thread and giving other coroutines a chance to run on it.
Suspending the coroutine lets it check for cancellation and throw `CancellationException` if it's canceled.

So `yield()` has two effects: it lets other coroutines run and checks for cancellation.

Here's an example:

```kotlin
import kotlinx.coroutines.*

fun main() {
//sampleStart
// runBlocking uses the current thread for running all coroutines
runBlocking {
    val coroutineCount = 5
    repeat(coroutineCount) { coroutineIndex ->
        launch {
            val id = coroutineIndex + 1
            repeat(5) { iterationIndex ->
                val iteration = iterationIndex + 1

>                 // Temporarily suspends to give other coroutines a chance to run

Reviewer: ```suggestion
                // Suspends temporarily to give other coroutines a chance to run
```

                // Without this, the coroutines run sequentially
                yield()
                // Prints the coroutine index and iteration index
                println("$id * $iteration = ${id * iteration}")
            }
        }
    }
}
//sampleEnd
}
```
{kotlin-runnable="true" id="yield-example"}

> In this example, each coroutine uses `yield()` to let other coroutines run between iterations.

Reviewer: ```suggestion
In this example, each coroutine uses the `yield()` function to let other coroutines run between iterations.
```

In this example, each coroutine uses `yield()` to let other coroutines run between iterations.

### Check for cancellation explicitly

> You almost never need to check for cancellation explicitly. In most cases, you can use `yield()`.

Reviewer: This is feels really vague 🤔 I understand the intention, but this needs more context. Why shouldn't the user check for it? What are the scenarios when the users might need it? 

Depending on what's your opinion on [this](https://github.com/Kotlin/kotlinx.coroutines/pull/4618/changes#r3520549879): I believe we should rewrite this section.

Alternatively, we could sidestep it, if we want to make it an official recommendation. 

> We don't recommend checking for cancellation explicitly unless...

 (if our stance is strong and we believe it warrants it)

Or something like: 
> Avoid checking for cancellation explicitly unless you need to check for cancellation without suspending.

Author: It is vague, yet explaining this will go into too much boring or complicated detail. Basically, use yield. It is prescribed as a spell here.

Reviewer: I'm not sure I understand the stance we are taking here.

What I'm saying is that our approach should depend on whether we want to officially recommend against it or not. If we have a (strong) opinion either way, we should state it explicitly. 

"You almost never need something" is a strange approach. There are many things we almost never need, yet we would like a description of how to do them.

Here's an example:

```kotlin
// Defines a coroutine scope that uses the UI thread
class ScreenWithButtons(private val scope: CoroutineScope) {

>     // Should only be called from the UI thread

Reviewer: general note: 
I suggest never using should. It's not exact, so it either lacks information or weakens the sentence. It should either explain why shouldn't the user do it, or if it doesn't matter, omit it.

    fun loadAndUpdateButtons(filename: String) {
        scope.launch {
            // withContext will check for cancellation before entering the block
            // and after the block returns
            val buttonNames = withContext(Dispatchers.IO) {

>                 readLines(filename) // A blocking call, cannot be canceled here

Reviewer: ```suggestion
                // This is a blocking call, and can't be canceled here
                readLines(filename)
```

            }

>             // If withContext() returned a value, it wasn't canceled.

Reviewer: This comment is very long. 😢  I'd suggest creating a description before the example.

            // Now this coroutine runs on the UI thread again,
            // so no one can cancel this component's scope and dispose of buttons
            // until this coroutine suspends for the next time and releases the UI thread.
            // It's safe to call updateUi here, as buttons are guaranteed to exist.
            //
            // In other words, if the screen is canceled while withContext() is running,
            // the coroutine throws CancellationException instead of assigning a value
            // to buttonNames, so updateUi() is never called.
            updateUi(buttonNames)
        }
    }

>     // Should only be called from the UI thread

Reviewer: ```suggestion
```

    // Throws an exception if called after the user left the screen
    private fun updateUi(buttonNames: List<String>) {
        // Updates buttons with the specified names.

>         // Throws if buttons no longer exist,

Reviewer: ```suggestion
        // Throws an exception if the buttons no longer exist
```

>         // due to being disposed of

Reviewer: ```suggestion
        // because they are disposed of after the screen closes
```

>         // after the user left the screen.

Reviewer: ```suggestion
```

    }

    // Should only be called from the UI thread
    fun leaveScreen() {
        // Cancels the scope when leaving the screen
        // You can no longer update the UI
        scope.cancel()
    }
}

// UI controller code
setHandler(Event.ScreenClosed) {

>     // Always executes from the UI thread

Reviewer: ```suggestion
    // Runs on the UI thread
```

        scope.cancel()
    }

> }

Reviewer 2: > `                readLines(filename) // suspend function, can be canceled`

That's probably wrong. In most cases, this exact code shape would contain a blocking call, because otherwise, why wouldn't `readLines` itself perform `withContext(Dispatchers.IO)` instead of requiring that from the caller?

In this specific scenario, the example would actually become weaker if `readLines` were a `suspend` function, because then, the race this section describes would trigger less often.

> `            // If the screen is cancelled while withContext() is executing,`

This is restating the same thing twice in a contrapositive. That can be helpful if you preface the second part with `In other words, ` or something like that, so that people know not to look for new information there.

Other than that, thanks, this is an improvement w.r.t. clarity!

> }

Author: Edited that, and added some more hints.
